mailet/standard/src/main/java/org/apache/james/transport/mailets/LogMessage.java

```diff
@@ -100,25 +131,55 @@ public void service(Mail mail) {
         }
     }
 
+    private void logAttribute(Mail mail) {
+        if (!specificAttributeNames.isEmpty()) {
+            logBuffer.get().append('\n');
+            for (String name : specificAttributeNames) {
+                AttributeName attributeName = AttributeName.of(name);
+                Optional<Attribute> attribute = mail.getAttribute(attributeName);
+                if (attribute.isPresent()) {
+                    logBuffer.get().append(name + ": " + attribute.get().getValue().value() + '\n');
+                } else {
+                    logBuffer.get().append(name + ": \n");
+                }
+            }
+        }
+    }
+
     private void logComment() {
         if (comment != null) {
-            logger.info(comment);
+            logBuffer.get().append(comment);
         }
     }
 
-    private void logHeaders(MimeMessage message) throws MessagingException {
+    private void logHeaders(Mail mail) throws MessagingException {
         if (headers && logger.isInfoEnabled()) {
-            logger.info("\n");
+            logBuffer.get().append('\n');
+            MimeMessage message = mail.getMessage();
             for (String header : Collections.list(message.getAllHeaderLines())) {
-                logger.info(header + "\n");
+                logBuffer.get().append(header + '\n');
+            }
+
+            if (!specificHeaderNames.isEmpty()) {
+                for (MailAddress recipient : mail.getRecipients()) {
+                    List<PerRecipientHeaders.Header> headers = mail.getPerRecipientSpecificHeaders()
+                            .getHeadersForRecipient(recipient).stream()
+                            .filter(header -> specificHeaderNames.contains(header.getName()))
+                            .collect(Collectors.toList());
+                    logBuffer.get().append("Recipient " + recipient.asPrettyString() + "'s headers are: \n");
+                    for (PerRecipientHeaders.Header header : headers) {
+                        logBuffer.get().append(header.getName() + ": " + header.getValue() + '\n');
```

Comment:
This, as all the for loops you introduced in this class could easily be replaced by a foreach.